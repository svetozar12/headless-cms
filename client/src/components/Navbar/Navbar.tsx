import * as React from "react";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";
import Menu from "@mui/material/Menu";
import MenuIcon from "@mui/icons-material/Menu";
import Container from "@mui/material/Container";
import Avatar from "@mui/material/Avatar";
import Button from "@mui/material/Button";
import Tooltip from "@mui/material/Tooltip";
import MenuItem from "@mui/material/MenuItem";
import AdbIcon from "@mui/icons-material/Adb";
import { signOut, useSession } from "next-auth/react";
import Router, { useRouter } from "next/router";
import { CONTENT_LIST, MODEL, MODEL_LIST } from "../../constants/routes";

const pages = [
  {
    title: "Model",
    path: MODEL_LIST,
    onClick: () => Router.push(MODEL_LIST),
  },
  {
    title: "Content",
    path: CONTENT_LIST,
    onClick: () => Router.push(CONTENT_LIST),
  },
];
const settings = [
  {
    label: "Logout",
    onClick: async () => {
      await signOut({ callbackUrl: "/" });
    },
  },
];

const Navbar = () => {
  const { data: session } = useSession();
  const { user } = session || {};
  const [anchorElNav, setAnchorElNav] = React.useState<null | HTMLElement>(
    null,
  );
  const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(
    null,
  );

  const handleOpenNavMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorElNav(event.currentTarget);
  };
  const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseNavMenu = () => {
    setAnchorElNav(null);
  };

  const handleCloseUserMenu = () => {
    setAnchorElUser(null);
  };

  if (!session) return <></>;

  return (
    <AppBar position="static">
      <Container maxWidth="xl">
        <Toolbar disableGutters>
          <Box sx={{ flexGrow: 1, display: { xs: "flex", md: "none" } }}>
            <IconButton
              size="large"
              aria-label="account of current user"
              aria-controls="menu-appbar"
              aria-haspopup="true"
              onClick={handleOpenNavMenu}
              color="inherit"
            >
              <MenuIcon />
            </IconButton>
            <Menu
              id="menu-appbar"
              anchorEl={anchorElNav}
              anchorOrigin={{
                vertical: "bottom",
                horizontal: "left",
              }}
              keepMounted
              transformOrigin={{
                vertical: "top",
                horizontal: "left",
              }}
              open={Boolean(anchorElNav)}
              onClose={handleCloseNavMenu}
              sx={{
                display: { xs: "block", md: "none" },
              }}
            >
              {pages.map(({ title }) => (
                <MenuItem key={title} onClick={handleCloseNavMenu}>
                  <Typography textAlign="center">{title}</Typography>
                </MenuItem>
              ))}
            </Menu>
          </Box>
          <AdbIcon sx={{ display: { xs: "flex", md: "none" }, mr: 1 }} />
          <Box sx={{ flexGrow: 1, display: { xs: "none", md: "flex" } }}>
            {pages.map(({ title, path, onClick }) => (
              <Button
                className={`${
                  window.location.pathname.includes(path) &&
                  "!bg-offBlack !bg-opacity-50"
                }`}
                key={title}
                onClick={() => {
                  handleCloseNavMenu();
                  onClick();
                }}
                sx={{ my: 2, color: "white", display: "block" }}
              >
                {title}
              </Button>
            ))}
          </Box>

          <Box sx={{ flexGrow: 0 }}>
            <Tooltip title="Open settings">
              <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                <Avatar
                  alt={user?.name as string}
                  src={user?.image as string}
                />
              </IconButton>
            </Tooltip>
            <Menu
              sx={{ mt: "45px" }}
              id="menu-appbar"
              anchorEl={anchorElUser}
              anchorOrigin={{
                vertical: "top",
                horizontal: "right",
              }}
              keepMounted
              transformOrigin={{
                vertical: "top",
                horizontal: "right",
              }}
              open={Boolean(anchorElUser)}
              onClose={handleCloseUserMenu}
            >
              {settings.map(({ label, onClick }) => (
                <MenuItem
                  key={label}
                  onClick={() => {
                    handleCloseUserMenu();
                    onClick();
                  }}
                >
                  <Typography textAlign="center">{label}</Typography>
                </MenuItem>
              ))}
            </Menu>
          </Box>
        </Toolbar>
      </Container>
    </AppBar>
  );
};
export default Navbar;
