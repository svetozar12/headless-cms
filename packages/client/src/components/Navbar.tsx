import useSession from "../hooks/useSession";
import Link from "next/link";
import { FaUserCircle } from "react-icons/fa";
import { MdDashboard, MdOutlineDashboard } from "react-icons/md";
import { AiFillHome, AiOutlineLogout, AiOutlineHome } from "react-icons/ai";
import { SiCraftcms } from "react-icons/si";
import { destroyCookie, parseCookies } from "nookies";
import { useEffect, useState } from "react";

const Navbar = () => {
  const { isLoggedIn } = useSession();
  const [activeTab, setActiveTab] = useState("home");
  const cookie = parseCookies();
  console.log(isLoggedIn);
  if (!isLoggedIn) return null;

  const logout = () => {
    for (const key in cookie) destroyCookie(null, key);
    window.location.reload();
  };

  useEffect(() => {
    if (window.location.pathname === "/") setActiveTab("home");
    else if (window.location.pathname === "/dashboard")
      setActiveTab("dashboard");
    else if (window.location.pathname === "/profile") setActiveTab("profile");
  });

  return (
    <nav className="flex h-14 w-full items-center  bg-mainPurple">
      <div className="navbar__logo h-full">
        <Link href="/">
          <SiCraftcms size="2.5rem" className="cursor-pointer" />
        </Link>
      </div>
      <div className="flex h-full w-full items-center justify-center gap-5">
        <Link href="/">
          {activeTab === "home" ? (
            <AiFillHome
              size="1.75rem"
              className="cursor-pointer"
              onClick={() => setActiveTab("home")}
            />
          ) : (
            <AiOutlineHome
              size="1.75rem"
              className="cursor-pointer"
              onClick={() => setActiveTab("home")}
            />
          )}
        </Link>
        <Link href="/dashboard">
          {activeTab === "dashboard" ? (
            <MdDashboard size="1.75rem" className="cursor-pointer" />
          ) : (
            <MdOutlineDashboard size="1.75rem" className="cursor-pointer" />
          )}
        </Link>
        <Link href="/profile">
          <FaUserCircle size="1.75rem" className="cursor-pointer" />
        </Link>
        <button type="button" onClick={logout}>
          <AiOutlineLogout size="1.75rem" className="cursor-pointer" />
        </button>
      </div>
    </nav>
  );
};

export default Navbar;
