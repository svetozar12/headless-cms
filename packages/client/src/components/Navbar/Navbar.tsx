import Link from "next/link";
import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
import { FaBoxes } from "react-icons/fa";
import { MdContentPaste } from "react-icons/md";
import {
  CONTENT,
  CONTENT_MODELS,
  LOGOUT,
  PROFILE,
} from "../../constants/routes";
import useSession from "../../hooks/useSession";
import { useCookie } from "next-cookie";
import Image from "next/image";
import Dropdown from "./subcomponents/Dropdown";
import Spinner from "../Spinner";

const HrefToTab = (href: string) => {
  return href.slice(1, href.length);
};

const Navbar = () => {
  const cookie = useCookie();
  const { user } = useSession();
  const { activeTab = CONTENT_MODELS } = useActiveTab();
  const router = useRouter();
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const NavLinks = [
    { Icon: FaBoxes, href: CONTENT_MODELS },
    {
      Icon: MdContentPaste,
      href: CONTENT,
    },
  ];

  const render = () => {
    return (
      <nav className="relative flex h-20 w-full items-center bg-mainPurple">
        {!user && <Spinner isLoading={!user} />}
        <div className="flex h-full w-full items-center justify-center gap-5">
          {NavLinks.map(({ Icon, href }) => {
            const tabName = HrefToTab(href);
            const isActiveTab = tabName === activeTab;

            return (
              <div
                key={href}
                className={`flex h-12 w-12 items-center justify-center rounded-md duration-500 ease-out bg-transparent${
                  isActiveTab && "rounded-md bg-textPurple"
                }`}
              >
                <Link key={href} href={href}>
                  <Icon
                    size="2.3rem"
                    className="cursor-pointer"
                    onClick={() => {
                      router.push(href);
                    }}
                  />
                </Link>
              </div>
            );
          })}

          <div className="relative h-9 w-9">
            <Image
              src={user?.avatar as any}
              width="37px"
              height="37px"
              className="flex cursor-pointer items-center justify-center rounded-full"
              onClick={() => setIsOpen((prev) => !prev)}
            />
            <div className="relative flex h-full w-full">
              <Dropdown
                isOpen={isOpen}
                items={[
                  {
                    title: "Logout",
                    onClick: () => {
                      router.push(LOGOUT);
                      setIsOpen(false);
                    },
                  },
                ]}
              />
            </div>
          </div>
        </div>
      </nav>
    );
  };

  return <>{cookie.get("accessToken") && render()}</>;
};

export default Navbar;

const useActiveTab = () => {
  const [activeTab, setActiveTab] = useState("");
  const { setTokens } = useSession();
  useEffect(() => {
    setTokens();

    if (window.location.pathname === CONTENT_MODELS)
      setActiveTab(() => HrefToTab(CONTENT_MODELS));
    else if (window.location.pathname === CONTENT)
      setActiveTab(() => HrefToTab(CONTENT));
    else if (window.location.pathname === PROFILE)
      setActiveTab(() => HrefToTab(PROFILE));
  }, [activeTab, window.location.pathname]);

  return { activeTab };
};
