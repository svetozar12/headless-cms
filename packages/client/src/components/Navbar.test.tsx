import { beforeEach, describe, expect, it, vi } from "vitest";
import { mount, ReactWrapper } from "enzyme";
import Navbar from "./Navbar";

describe("Navbar component", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(<Navbar />);
  });
  it("should render navbar while logged", () => {
    vi.mock("../hooks/useSession", async () => {
      const originalModule: any = await vi.importActual("../hooks/useSession");
      return {
        __esModule: true,
        ...originalModule,
        default: () => ({
          isLoggedIn: true,
          user: { username: "test" },
        }),
      };
    });
    expect(wrapper.find("nav").length).toBe(1);
    expect(wrapper.find("a").length).toBe(2);
    expect(wrapper.find("a").at(0).text()).toBe("Dashboard");
    expect(wrapper.find("a").at(1).text()).toBe("Profile");
    expect(wrapper.find("button").text()).toBe("LogOut");
    expect(wrapper).toMatchSnapshot();
  });
});
