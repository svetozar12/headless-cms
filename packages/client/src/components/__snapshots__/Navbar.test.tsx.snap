// Vitest Snapshot v1

exports[`Navbar component > should render navbar 1`] = `ReactWrapper {}`;

exports[`Navbar component > should render navbar while logged 1`] = `ReactWrapper {}`;
